pkg/graveler/committed/diff.go

```diff
@@ -112,61 +170,91 @@ func (d *diffIterator) Next() bool {
+			return true
+		}
+		d.emptyCurrentRange()
+	}
 	select {
 	case <-d.ctx.Done():
 		d.err = d.ctx.Err()
 		return false
 	default:
-	}
-	if d.state == diffIteratorStatePreInit {
-		d.state = diffIteratorStateOpen
-		d.leftValue.record, d.leftValue.rng, d.leftValue.err = diffIteratorNextValue(d.left)
-		d.rightValue.record, d.rightValue.rng, d.rightValue.err = diffIteratorNextValue(d.right)
-	}
-	for {
-		if d.rightValue.err != nil {
-			d.err = d.rightValue.err
-		}
-		if d.leftValue.err != nil {
-			d.err = d.leftValue.err
-		}
-		if d.err != nil {
-			d.currentDiff = nil
-			return false
-		}
-		compareResult := d.compareDiffIterators()
-		switch compareResult {
-		case diffItCompareResultDone:
-			d.currentDiff = nil
-			return false
-		case diffItCompareResultSameRanges:
-			d.leftValue.rng, d.leftValue.err = diffIteratorNextRange(d.left)
-			d.rightValue.rng, d.rightValue.err = diffIteratorNextRange(d.right)
-		case diffItCompareResultSameKeys:
-			// same keys on different ranges
-			d.currentDiff = &graveler.Diff{Type: graveler.DiffTypeChanged, Key: d.rightValue.record.Key.Copy(), Value: d.rightValue.record.Value, LeftIdentity: d.leftValue.record.Identity}
-			d.leftValue.record, d.leftValue.rng, d.leftValue.err = diffIteratorNextValue(d.left)
-			d.rightValue.record, d.rightValue.rng, d.rightValue.err = diffIteratorNextValue(d.right)
-			return true
-		case diffItCompareResultSameIdentities, diffItCompareResultNeedStartRangeBoth:
-			d.leftValue.record, d.leftValue.rng, d.leftValue.err = diffIteratorNextValue(d.left)
-			d.rightValue.record, d.rightValue.rng, d.rightValue.err = diffIteratorNextValue(d.right)
-		case diffItCompareResultNeedStartRangeLeft:
+		if d.state == diffIteratorStatePreInit {
+			d.state = diffIteratorStateOpen
 			d.leftValue.record, d.leftValue.rng, d.leftValue.err = diffIteratorNextValue(d.left)
-		case diffItCompareResultNeedStartRangeRight:
 			d.rightValue.record, d.rightValue.rng, d.rightValue.err = diffIteratorNextValue(d.right)
-		case diffItCompareResultLeftBeforeRight:
-			// nothing on right, or left before right
-			d.currentDiff = &graveler.Diff{Type: graveler.DiffTypeRemoved, Key: d.leftValue.record.Key.Copy(), Value: d.leftValue.record.Value, LeftIdentity: d.leftValue.record.Identity}
-			d.leftValue.record, d.leftValue.rng, d.leftValue.err = diffIteratorNextValue(d.left)
-			return true
-		case diffItCompareResultRightBeforeLeft:
-			// nothing on left, or right before left
-			d.currentDiff = &graveler.Diff{Type: graveler.DiffTypeAdded, Key: d.rightValue.record.Key.Copy(), Value: d.rightValue.record.Value}
-			d.rightValue.record, d.rightValue.rng, d.rightValue.err = diffIteratorNextValue(d.right)
-			return true
 		}
+		for {
+			if d.rightValue.err != nil {
+				d.err = d.rightValue.err
+			}
+			if d.leftValue.err != nil {
+				d.err = d.leftValue.err
+			}
+			if d.err != nil {
+				d.currentDiff = nil
+				d.emptyCurrentRange()
```

Comment:
True, but we return nil in values in case of errors 
```// In case `Next()` returns false, `Value()` returns nil and `Err()` should be checked.```
At least this is what we wrote in the documentation in graveler